config.go

```diff
@@ -6,53 +6,130 @@ import (
 	"net/http"
 	"net/url"
 	"os"
+	"reflect"
 )
 
 type Opt interface {
 	apply(*cfg)
+	Type() string
 }
 
 type cfg struct {
 	controllerUrl  string
 	authToken      string
+	authType       string
 	httpAuthWriter httpAuthWriter
 	logger         *slog.Logger
 }
 
 type clientOpt struct{ fn func(*cfg) }
 
+// controllerUrlOpt is an option to set the controller url for the client
+type controllerUrlOpt struct {
+	controllerUrl string
+}
+
+func (o *controllerUrlOpt) apply(c *cfg) {
+	c.controllerUrl = o.controllerUrl
+}
+
+func (o *controllerUrlOpt) Type() string {
+	return "controllerUrl"
+}
+
+// authTokenOpt is an option to set the auth token for the client
+type authTokenOpt struct {
+	authToken string
+}
+
+func (o *authTokenOpt) apply(c *cfg) {
+	c.authToken = o.authToken
+}
+
+func (o *authTokenOpt) Type() string {
+	return "authToken"
+}
+
+// authTypeOpt is an option to set the auth type for the client
+type authTypeOpt struct {
+	authType string
+}
+
+func (o *authTypeOpt) apply(c *cfg) {
+	c.authType = o.authType
+}
+
+func (o *authTypeOpt) Type() string {
+	return "authType"
+}
+
+// loggerOpt is an option to set the logger for the client
+type loggerOpt struct {
+	logger *slog.Logger
+}
+
+func (o *loggerOpt) apply(c *cfg) {
+	c.logger = o.logger
+}
+
+func (o *loggerOpt) Type() string {
+	return "logger"
+}
+
 func (opt clientOpt) apply(cfg *cfg) { opt.fn(cfg) }
 
 func ControllerUrl(pinotControllerUrl string) Opt {
-	return clientOpt{fn: func(cfg *cfg) { cfg.controllerUrl = pinotControllerUrl }}
+	return &controllerUrlOpt{controllerUrl: pinotControllerUrl}
 }
 
 func AuthToken(token string) Opt {
-	return clientOpt{fn: func(cfg *cfg) { cfg.authToken = token }}
+	return &authTokenOpt{authToken: token}
 }
 
 func Logger(logger *slog.Logger) Opt {
-	return clientOpt{fn: func(cfg *cfg) { cfg.logger = logger }}
+	return &loggerOpt{logger: logger}
+}
+
+func AuthType(authType string) Opt {
+	return &authTypeOpt{authType: authType}
 }
 
 func validateOpts(opts ...Opt) (*cfg, *url.URL, error) {
 
 	// with default auth writer that does nothing
 	optCfg := defaultCfg()
+	optCounts := make(map[string]int)
 	for _, opt := range opts {
+		optType := reflect.TypeOf(opt).Elem().Name()
```

Comment:
its also possible to avoid implementing `Type()`

```go
	switch v := i.(type) {
	case int:
		fmt.Printf("Twice %v is %v\n", v, v*2)
	case string:
		fmt.Printf("%q is %v bytes long\n", v, len(v))
	default:
		fmt.Printf("I don't know about type %T!\n", v)
	}
```

not the biggest fan of using reflection

Reply:
yeah smoother, see new commit

Reply:
keeping type just for default case